sklearn/utils/stats.py

```diff
@@ -102,22 +94,51 @@ def _weighted_percentile(array, sample_weight, percentile_rank=50, xp=None):
         ],
         device=device,
     )
-    # In rare cases, `percentile_indices` equals to `sorted_idx.shape[0]`
     max_idx = sorted_idx.shape[0] - 1
     percentile_indices = xp.clip(percentile_indices, 0, max_idx)
-
     col_indices = xp.arange(array.shape[1], device=device)
     percentile_in_sorted = sorted_idx[percentile_indices, col_indices]
-
     result = array[percentile_in_sorted, col_indices]
 
+    if symmetrize:
+        # Use reverse of the original sorted order for the negated array
+        sorted_idx_neg = xp.flip(sorted_idx, axis=0)
+        sorted_weights_neg = xp.take_along_axis(sample_weight, sorted_idx_neg, axis=0)
+        array_neg = -array
+        largest_value_neg = array_neg[
+            sorted_idx_neg[-1, ...], xp.arange(n_features, device=device)
+        ]
+        if xp.any(xp.isnan(largest_value_neg)):
+            sorted_nan_mask_neg = xp.take_along_axis(
+                xp.isnan(array_neg), sorted_idx_neg, axis=0
+            )
+            sorted_weights_neg[sorted_nan_mask_neg] = 0
+
+        weight_cdf_neg = xp.cumulative_sum(sorted_weights_neg.T, axis=1)
+        adjusted_pr_neg = (100 - percentile_rank) / 100 * weight_cdf_neg[..., -1]
+        mask_neg = adjusted_pr_neg == 0
+        adjusted_pr_neg[mask_neg] = xp.nextafter(
+            adjusted_pr_neg[mask_neg], adjusted_pr_neg[mask_neg] + 1
+        )
+        percentile_indices_neg = xp.asarray(
+            [
+                xp.searchsorted(
+                    weight_cdf_neg[feature_idx, ...], adjusted_pr_neg[feature_idx]
+                )
+                for feature_idx in range(weight_cdf_neg.shape[0])
+            ],
+            device=device,
+        )
+        percentile_indices_neg = xp.clip(percentile_indices_neg, 0, max_idx)
+        percentile_in_sorted_neg = sorted_idx_neg[percentile_indices_neg, col_indices]
+        result_neg = array_neg[percentile_in_sorted_neg, col_indices]
+
+        result = (result - result_neg) / 2
+
     return result[0] if n_dim == 1 else result
 
 
-# TODO: refactor to do the symmetrisation inside _weighted_percentile to avoid
-# sorting the input array twice.
 def _averaged_weighted_percentile(array, sample_weight, percentile_rank=50, xp=None):
```

Comment:
Why not just replace instances of `_averaged_weighted_percentile` with `_weighted_percentile(symmetrize=True)` ?